File: common/src/main/java/io/xol/chunkstories/entity/SerializedEntityFile.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.entity;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

import io.xol.chunkstories.api.entity.Entity;
import io.xol.chunkstories.api.world.World;
import io.xol.chunkstories.api.world.serialization.OfflineSerializedData;

public class SerializedEntityFile implements OfflineSerializedData {
	private final File file;

	public SerializedEntityFile(String string) {
		file = new File(string);
	}

	public boolean exists() {
		return file.exists();
	}

	public String toString() {
		return "[CSF File: " + file.toString() + "]";
	}

	public Entity read(World world) {
		try {
			DataInputStream in = new DataInputStream(new FileInputStream(file));

			Entity entity = EntitySerializer.readEntityFromStream(in, this, world);

			in.close();

			return entity;
		} catch (IOException e) {
			e.printStackTrace();
		}
		return null;
	}

	public void write(Entity entity) {
		try {
			file.getParentFile().mkdirs();

			DataOutputStream out = new DataOutputStream(new FileOutputStream(file));

			EntitySerializer.writeEntityToStream(out, this, entity);

			out.flush();
			out.close();

			System.out.println("Wrote serialized entity to : " + file);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: common/src/main/java/xyz/chunkstories/entity/SerializedEntityFile.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.entity

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.world.World
import xyz.chunkstories.api.world.serialization.OfflineSerializedData

class SerializedEntityFile(string: String) : OfflineSerializedData {
    private val file: File

    init {
        file = File(string)
    }

    fun exists(): Boolean {
        return file.exists()
    }

    override fun toString(): String {
        return "[CSF File: $file]"
    }

    fun read(world: World): Entity? {
        try {
            val dis = DataInputStream(FileInputStream(file))

            val entity = EntitySerializerOld.readEntityFromStream(dis, this, world)

            dis.close()

            return entity
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    fun write(entity: Entity) {
        try {
            file.parentFile.mkdirs()

            val out = DataOutputStream(FileOutputStream(file))

            EntitySerializerOld.writeEntityToStream(out, this, entity)

            out.flush()
            out.close()

            println("Wrote serialized entity to : $file")
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }
}
